text/0000-futures.md

```diff
@@ -0,0 +1,520 @@
+[guide-level-explanation]: #guide-level-explanation
+
+The `Future` trait represents an *asynchronous* and lazy computation that may
+eventually produce a final value, but doesn't have to block the current thread
+to do so.
+
+Futures can be constructed through `async` blocks or `async` functions, e.g.,
+
+```rust
+async fn read_frame(socket: &TcpStream) -> Result<Frame, io::Error> { ... }
+```
+
+This `async` function, when invoked, produces a future that represents the
+completion of reading a frame from the given socket. The function signature
+is equivalent to:
+
+```rust
+fn read_frame<'sock>(socket: &'sock TcpStream)
+    -> impl Future<Output = Result<Frame, io::Error>> + 'sock;
+```
+
+Other async functions can *await* this future; see the [companion
+RFC] for full details.
+
+In addition to `async fn` definitions, futures can be built using adapters, much
+like with `Iterator`s. Initially these adapters will be provided entirely "out
+of tree", but eventually they will make their way into the standard library.
+
+Ultimately asynchronous computations are executed by *tasks*, which are
+lightweight threads. In particular, an *executor* is able to "spawn" a
+`()`-producing `Future` as an independent task; these tasks are then
+cooperatively scheduled onto one or more operating system threads. The
+`Executor` trait defines this interface, and the `task` module provides a host
+of related definitions needed when manually implementing `Future`s or
+executors.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## `core::task` module
+
+The fundamental mechanism for asynchronous computation in Rust is *tasks*, which
+are lightweight threads of execution; many tasks can be cooperatively scheduled
+onto a single operating system thread.
+
+To perform this cooperative scheduling we use a technique sometimes referred to
+as a "trampoline". When a task would otherwise need to block waiting for some
+event, instead it schedules itself for later wakeup and *returns* to the
+executor running it, which can then run another task. Subsequent wakeups place
+the task back on the executors queue of ready tasks, much like a thread
+scheduler in an operating system.
+
+Attempting to complete a task (or async value within it) is called *polling*,
+and always yields a `Poll` value back:
+
+```rust
+/// Indicates whether a value is available, or if the current task has been
+/// scheduled for later wake-up instead.
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub enum Poll<T> {
+    /// Represents that a value is immediately ready.
+    Ready(T),
+
+    /// Represents that a value is not ready yet.
+    ///
+    /// When a function returns `Pending`, the function *must* also
+    /// ensure that the current task is scheduled to be awoken when
+    /// progress can be made.
+    Pending,
+}
+```
+
+When a task returns `Poll::Ready`, the executor knows the task has completed and
+can be dropped.
+
+### Waking up
```

Comment:
It's unclear at first glance which of the code blocks starting with trait Wake/struct ExecutorInner/struct Waker are proposed for stabilization.